miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/RealMiniAppDisplay.kt

```diff
@@ -0,0 +1,45 @@
+package com.rakuten.tech.mobile.miniapp.display
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.view.View
+import android.view.ViewGroup
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import android.widget.FrameLayout
+import androidx.annotation.VisibleForTesting
+import com.rakuten.tech.mobile.miniapp.MiniAppDisplay
+
+@SuppressLint("SetJavaScriptEnabled")
+internal class RealMiniAppDisplay(context: Context, val basePath: String) : MiniAppDisplay,
+    WebView(context) {
+
+    init {
+        setLayoutParams(
+            FrameLayout.LayoutParams(
+                ViewGroup.LayoutParams.MATCH_PARENT,
+                ViewGroup.LayoutParams.MATCH_PARENT
+            )
+        )
+        settings.javaScriptEnabled = true
+        settings.allowUniversalAccessFromFileURLs = true
+        webViewClient = MiniAppWebViewClient()
+        loadUrl(getLoadUrl())
+    }
+
+    override fun setWebViewClient(client: WebViewClient?) {
+        super.setWebViewClient(client ?: MiniAppWebViewClient())
+    }
+
+    @VisibleForTesting
+    fun getLoadUrl() = "file://$basePath/index.html"
```

Comment:
Looks like this function could be internal. 

Or can't we test the load url by directly calling `WebView#url`? And then this function can be private.

Reply:
Actually, there are two ways to do it.
1) Functional way: Try to break down the units into the smallest pieces. And then, try to test the smaller units.
2) Without the explicit method: The string expression can directly be embedded into loadUrl. This can be unit tested by WebView#url.

Currently, I am trying to follow the first approach. Making it internal is not bringing any benefit as such because the class is already internal and can not be used for typecasting. 💭 🤔 💭

Reply:
Yeah, I agree that things can be tested in a functional way, but it doesn't make sense to do so when the function is exposed only so that it can be tested. In that case, it's the same thing as testing internal implementation details.

My point is that the reason for testing this is to make sure the WebView is loading the correct URL, but just testing an exposed `getLoadUrl` function gives us no guarantee that the WebView is actually using that function and loading the correct URL.

Reply:
Thanks for sharing your thoughts.

Actually, the intent is to just test if the URL forming logic is correct or not (which can be an easy source of error otherwise). I had marked it `VisibleForTesting`. 
To be in parity with `RetrofitCreatorUtils`, I have updated the modifiers.

Reply:
You're right, we do need to test that the URL was formed correctly, but I don't it should be tested by exposing a function. The reason we need to test if the URL was formed correctly is to check if the WebView is displaying the correct URL. But if we were to change the call for `WebView#loadUrl` and are no longer calling the `getLoadUrl` function, then the test for `getLoadUrl` would be meaningless.

Actually, I think it would be better if `"file://$basePath/index.html"` were moved inline to `WebView#loadUrl` instead of being returned by a separate function. `getLoadUrl` is only being used by the WebView and it doesn't need to be known by anything else.

Sorry I keep coming back to this issue, but I think it's important that we avoid this kind of pattern where some functionality is exposed only so that it can be used in unit tests. It can make the code more difficult to read and can be a potential source of bugs later on.

Reply:
Your description has two different contexts, I will discuss this with you directly. Unit testing a block of logic code is different from unit testing its usage by its consumer.